### PR TITLE
[Remote Store] Avoiding Translog Deletion in case of relocating shards. 

### LINKED ISSUE #9191: [BUG] org.opensearch.remotestore.RemoteIndexPrimaryRelocationIT.testPrimaryRelocationWhileIndexing flaky failure
**Describe the bug**

4 failures in org.opensearch.remotestore.RemoteIndexPrimaryRelocationIT.testPrimaryRelocationWhileIndexing (21818,21853,22063,22065)

Meta issue - https://github.com/opensearch-project/OpenSearch/issues/8279

### PR DESCRIPTION
**Pull request by gbbafna**
### Description

In Primary to Primary relocation , there can be concurrent upload and download of translog.
While translog files are getting downloaded by new primary, it might hence be deleted by the relocating primary
Hence we need to avoid to deletion when shard is in `RELOCATING` state . 

### Related Issues

Resolves #9191

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteIndexPrimaryRelocationIT.java [hunk 1]
```diff
@@ -64,7 +64,6 @@ protected Settings featureFlagSettings() {
             .build();
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/opensearch-project/OpenSearch/issues/9191")
     public void testPrimaryRelocationWhileIndexing() throws Exception {
```
Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteIndexPrimaryRelocationIT.java:67]: Lets remove the override now.

Comment (ashking94) [inline server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteIndexPrimaryRelocationIT.java:67]: Lets remove the override now.

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -337,6 +338,29 @@ Runnable getGlobalCheckpointSyncer() {
 
     private final List<ReferenceManager.RefreshListener> internalRefreshListener = new ArrayList<>();
 
+    /**
+     * An OpenSearch index shard config supplier
+     *
+     * @opensearch.internal
+     */
+    static public class IndexShardConfigSupplier {
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:346]: let's have a builder?

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java [hunk 3]
```diff
@@ -385,6 +388,10 @@ public void trimUnreferencedReaders() throws IOException {
             return;
         }
 
+        if (relocatingSupplier.getAsBoolean() == true) {
+            return;
+        }
+
```
Comment (ashking94) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:394]: Lets move this before the below code -
```
if (remoteGenerationDeletionPermits.tryAcquire(REMOTE_DELETION_PERMITS) == false) {
            return;
        }
```

**Review by ashking94 — CHANGES_REQUESTED** [review]
Looks good overall, lets add UTs/ITs. Thanks for the change.

File: server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteIndexPrimaryRelocationIT.java (hunk 1 above)
Comment (gbbafna) [reply server/src/internalClusterTest/java/org/opensearch/remotestore/RemoteIndexPrimaryRelocationIT.java:67]: done.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java (hunk 3 above)
Comment (gbbafna) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java:394]: good catch. This permits has made the code prone to errors.

**Comment by gbbafna** [comment]
Failure not related to this change : https://github.com/opensearch-project/OpenSearch/issues/9580

**Review by Bukhtawar — COMMENTED** [review]
What if relocation is stuck and not all readers are needed, would we end up retaining readers for long? I would model this as a retention lease instead, the interface looks generic and cleaner that way

File: server/src/main/java/org/opensearch/index/shard/IndexShard.java (hunk 2 above)
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/shard/IndexShard.java:346]: The structure seems to be force-fitting this very specific use case. We should model it more generically

**Pull request closed without merging**